File: CrmApp.Client/src/models/ContractorContracts.ts
import { ContractorHoursSnapshots } from "./ContractorHoursSnapshots";
import { EBillingType } from "./enums/EBillingType";
import { EEngagementType } from "./enums/EEngagementType";
import { ERenewalType } from "./enums/ERenewalType";
import { UsersProfiles } from "./UsersProfiles";

export class ContractorContracts {
    id: number = 0;
    contractorId: number = 0;
    contractNumber: string = "";
    engagementType: EEngagementType = EEngagementType.None;
    createdAt: Date = new Date();
    createdBy: number = 0;
    validFrom?: Date;
    validTo?: Date;
    hoursLimit: number = 0;
    modifiedBy?: number;
    modifiedAt?: Date;
    billingType: EBillingType = EBillingType.None;
    billingAmount?: number;
    allowOverLimit: boolean = false;
    renewalType: ERenewalType = ERenewalType.None;
    renewalDate?: Date;

    engagementTypeName: string = "";
    hoursUsed?: number = 0;
    hoursRemaining?: number = 0;
    hoursUsedCurrentMonth?: number = 0;
    hoursRemainingCurrentMonth?: number = 0;
    hoursFromPrevMonth?: number = 0;
    snapshots: ContractorHoursSnapshots[] = [];
    validFromStr?: string;
    validToStr?: string;
    renewalDateStr?: string;

    setDefaultBillingTypeByEngagementType() {
        switch (this.engagementType) {
            case EEngagementType.None:
                this.billingType = EBillingType.None;
                break;
            case EEngagementType.Contract:
                this.billingType = EBillingType.OneTimePayment;
                break;
            case EEngagementType.HoursPackage:
                this.billingType = EBillingType.Monthly;
                break;
            case EEngagementType.MonthlyBilling:
                this.billingType = EBillingType.Monthly;
                break;
            default:
                this.billingType = EBillingType.None;
        }   
    }
}

File: CrmApp.Domain/Entities/ContractorContracts.cs
using CrmApp.Domain.DTO;
using CrmApp.Domain.Enums;
using CrmApp.Domain.Tools;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CrmApp.Domain.Entities
{
    public class ContractorContracts : ExtendableClass
    {
        [Key]
        public int Id { get; set; }
        public int ContractorId { get; set; }
        public string ContractNumber { get; set; } = string.Empty;
        public EEngagementType EngagementType { get; set; }
        public DateTime CreatedAt { get; set; }
        public int CreatedBy { get; set; }
        public DateTime? ValidFrom { get; set; }
        public DateTime? ValidTo { get; set; }
        public decimal HoursLimit { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedAt { get; set; }
        public EBillingType BillingType { get; set; }
        public decimal? BillingAmount { get; set; }
        public bool AllowOverLimit { get; set; }
        public ERenewalType RenewalType { get; set; }
        public DateTime? RenewalDate { get; set; }

        [ForeignKey(nameof(ContractorId))]
        public virtual Contractors? Contractor { get; set; }
        [ForeignKey(nameof(CreatedBy))]
        public virtual UsersProfiles? CreatedByUser { get; set; }

        [NotMapped] public string EngagementTypeName => EngagementType.GetDescription();
        [NotMapped] public decimal? HoursUsed { get; set; }
        [NotMapped] public decimal? HoursRemaining { get; set; }
        [NotMapped] public string? ValidFromStr { get; set; }
        [NotMapped] public string? ValidToStr { get; set; }
        [NotMapped] public string? RenewalDateStr { get; set; }
        [NotMapped] public decimal? HoursUsedCurrentMonth { get; set; }
        [NotMapped] public decimal? HoursRemainingCurrentMonth { get; set; }
        [NotMapped] public decimal? HoursFromPrevMonth { get; set; }
        [NotMapped] public IReadOnlyList<ContractorHoursSnapshots> Snapshots { get; set; } = [];
    }
}
